frontend/lib/resolution/try-catch-analysis.cpp

```diff
@@ -0,0 +1,376 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "try-catch-analysis.h"
+
+#include "chpl/parsing/parsing-queries.h"
+#include "chpl/resolution/ResolvedVisitor.h"
+#include "chpl/resolution/resolution-types.h"
+#include "chpl/resolution/scope-queries.h"
+#include "chpl/uast/all-uast.h"
+#include "chpl/framework/compiler-configuration.h"
+
+#include "Resolver.h"
+
+
+namespace chpl {
+namespace resolution {
+
+
+using namespace uast;
+using namespace types;
+
+
+typedef enum {
+  ERROR_MODE_UNKNOWN,
+  ERROR_MODE_FATAL,
+  ERROR_MODE_RELAXED,
+} errorCheckingMode;
+
+  struct TryCatchState {
+    const AstNode* ast;
+    int  tryDepth;
+    bool fnDoesThrow;
+    int taskFunctionDepth;
+  };
+
+  struct TryCatchAnalyzer {
+    using RV = ResolvedVisitor<TryCatchAnalyzer>;
+
+    // input
+    Context* context;
+    const AstNode* astForErr;
+    errorCheckingMode mode;
+    // need to keep a stack of current ast and some meta info
+    // like if we're in a place it's ok to throw, what the current
+    // error checking mode is, etc.
+    std::vector<TryCatchState> tryStack;
+    int tryDepth = 0;
+    // output
+    // TODO: figure out the output - errors when something doesn't pass?
+    // maybe no output?
+    std::vector<QualifiedType> returnedTypes;
+
+    TryCatchAnalyzer(Context* context,
+                  const AstNode* symbol,
+                  errorCheckingMode mode)
+        : context(context),
+          astForErr(symbol),
+          mode(mode) {
+    }
+
+    void checkCatches(const Try* node, RV& rv) {
+
+      bool hasCatchAll = false;
+      for (auto handler : node->handlers()) {
+        if (hasCatchAll) {
+          context->error(node, "catchall placed before the end of a catch list");
+        }
+        if (isCatchAll(handler, rv)) {
+          hasCatchAll = true;
+        }
+      }
+    }
+
+    // Returns true if the catch is a catchall statement.
+    bool isCatchAll(const Catch* node, RV& rv) {
+        bool catchAll = false;
+        if (auto err = node->error()) {
+          // we validate that the type of error is a subtype of Error or Error
+          // itself resolving the AST.
+          // So here we just need to check for the base Errors.Error to know if
+          // this is a catchall.
+          const ResolvedExpression& r = rv.byAst(err);
+          auto errType = CompositeType::getErrorType(context);
+          auto dec = ClassTypeDecorator(ClassTypeDecorator::MANAGED_NONNIL);
+          auto manager = AnyOwnedType::get(context);
+          auto ret = ClassType::get(context, errType->basicClassType(),
+                                    manager, dec);
+          auto qt = QualifiedType(QualifiedType::VAR,
+                                  ret->withDecorator(context, dec));
+          if (r.type() == qt) {
+            catchAll = true;
+          }
```

Comment:
Instead of equality, would it be more resilient to use `canPass`, and ensure that it passes but doesn't convert via subtyping?